### PR TITLE
Add helper classes for shortcode testing

### LINKED ISSUE #12: Move `ShortcodeTest` base class somewhere else?
This bundle contains a [`ShortcodeTest`](https://github.com/webfactory/WebfactoryShortcodeBundle/blob/master/tests/Functional/ShortcodeTest.php
) base class supposed to help you writing tests for your own shortcodes.

There are a few issues with this class:

1. It is not covered by tests.
2. Were it covered by tests, we might have noticed that it refers to `PHPUnit_Framework_IncompleteTestError`, among others, and thus it only works with PHPUnit < 6.0.
3. It depends on `symfony/framework-bundle` for the `WebTestCase` base class, but that's not a general requirement of the bundle.
4. It is located in `tests/`, where it cannot easily be found by the composer autoloader _in other projects using this bundle here_.
5. If we move the class to `src/` as suggested in #10, `composer-require-checker` rightfully complains, because it then detects that we're using code from a package not declared as a dependency – see 2.

The only way I see right now to resolve the situation is to provide a separate Composer package, like `webfactory/shortcode-bundle-tests`. This package could contain the `ShortcodeTest` base class, but put it into the regular `autoload` Composer section. Additionally, it can declare supported – and at the same time, required – PHPUnit and Symfony FrameworkBundle versions in its `require` section.

Clients interested in using the test base class would `require-dev` this `webfactory/shortcode-bundle-tests` package. In `dev`, the PHPUnit dependency is not an issue and can be resolved as usual. Also, the `ShortcodeTest` class will be part of the regular autoloader configuration.

During `no-dev` installs, the entire package including the `ShortcodeTest` class would not even be installed.

### LINKED ISSUE #13: Additional/alternative solution to `ShortcodeTest`?
The `ShortcodeTest` base class is a full-fledged `WebTestCase`. It assumes that you have set up the [shortcode guide](https://github.com/webfactory/WebfactoryShortcodeBundle#activating-the-shortcode-guide), since it uses the rendering feature provided by that to display the shortcode under test in isolation.

That may be a bit complicated to use, for example when rendering the shortcode requires staging some database content. Imagine a shortcode that fetches some database rows and renders a table with it.

In such a case, you might not want to put different test data into the database, then run a `ShortcodeTest` with your shortcode and inspect the resulting HTML for correctness. It may be easier to directly test the controller generating the shortcode substitution HTML, or even the sub-modules or components that controller uses.

That leaves you with two types of tests:

1. For a given shortcode (name + optionally parameters), verify that the current configuration (DIC, tagged services etc.) would actually call the right controller with the right parameters when filtering content through `|shortcodes`; and that the HTML (or whatever response body returned) is actually embedded instead of the original shortcode
2. Test that the controllers generating shortcode substitution content work as expected for various datasets/inputs.

Regarding 1.), you could even argue that the substitution part is not worth testing if you rely on this bundle here doing its job properly. What you would like to assert as a user of this bundle is that you've done your configuration and service tagging correctly and that a shortcode you expect to work is registered and being processed.

(Note that even this test does not guarantee you use the `|shortcodes` filter where appropriate; this seems to be completely outside the scope of this bundle?)

**Bottom line:** Do we want to provide an additional/alternative test base class that supports testing the configuration works as expected, without actually running the controllers generating the shortcode substitution content, and without requiring the setup of the *shortcode guide* feature?

### LINKED ISSUE #24: Check shortcode definitions: controller resolvable?
It would be nice to have an easy way to test (either explicitly or implicitely) if the shortcode definitions are valid. The semantic configuration in #21 does a lot for that, but does not check whether a controller can be resolved.

`\Symfony\Component\HttpKernel\Controller\ControllerResolver` might help.

### PR DESCRIPTION
**Pull request by mpdude**
This PR addresses the shortcomings outlined in #12 and #13, and probably replaces #10 and #18 as well.

1. Deprecate the `\Webfactory\ShortcodeBundle\Tests\Functional\ShortcodeTest` class. 

* It was not available through the regular autoload mechanism (required extra trickery), 
* depended on undeclared dependencies (BrowserKit and DomCrawler), 
* came without tests, 
* suggested users to do/prefer full-scale Application Testing (using `WebTestCase`) and 
* worked only through the means of (ab-)using the Shortcode Guide.

2. Provide alternate means of testing shortcode processing

* Encourage users to do plain/direct unit testing (or integration testing, when necessary) of their shortcode handlers and/or shortcode handling controllers through new sections in the README. This gives best (most direct) control of input parameters.
* Add the public `\Webfactory\ShortcodeBundle\Test\ShortcodeDefinitionTestHelper` service that can be used in Kernel-based (functional) tests to verify a given shortcode name is known, the controller can be resolved etc.
* Add a `\Webfactory\ShortcodeBundle\Test\EndToEndTestHelper` class that simplifies full end-to-end (from content to processed content) functional testing, including round-trips through Symfony's FragmentHandler for controller-based shortcodes.

Fixes #12, fixes #13, closes #10, closes #18, closes #24.

### CODE REVIEW

File: tests/Functional/ShortcodeExtensionTest.php [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+/**
+ * Test that integration with Twig works as expected.
+ */
+class ShortcodeExtensionTest extends KernelTestCase
+{
+    /**
+     * @test
+     */
+    public function resolve__placeholder_shortcode_in__twig(): void
```
Comment (MalteWunsch) [inline tests/Functional/ShortcodeExtensionTest.php:15]: I've noticed these double underscores several times and wonder if I've missed some meaning?

File: tests/Fixtures/config/test_shortcodes.xml [hunk 2]
```diff
@@ -17,5 +17,9 @@
             <tag name="webfactory.shortcode" shortcode="test-service-inline" />
         </service>
 
+        <service id="Thunder\Shortcode\Handler\PlaceholderHandler">
+            <tag name="webfactory.shortcode" shortcode="placeholder" />
+        </service>
```
Comment (MalteWunsch) [inline tests/Fixtures/config/test_shortcodes.xml:22]: I'm not entirely happy for our tests to depend on this external PlaceholderHandler class. But it is officially mentioned in [thunderer's README](https://github.com/thunderer/Shortcode#handlers), so 👍

**Review by MalteWunsch — APPROVED** [review]
Good job 💚

File: tests/Functional/ShortcodeExtensionTest.php (hunk 1 above)
Comment (mpdude) [reply tests/Functional/ShortcodeExtensionTest.php:15]: No, somehow they keep reappearing when I run PHP-CS-Fixer.

**Pull request merged**